snuba/subscriptions/worker.py

```diff
@@ -35,21 +42,40 @@ def __init__(
 
     def process_message(
         self, message: Message[Tick]
-    ) -> Optional[Sequence[SubscriptionResult]]:
-        results: MutableSequence[SubscriptionResult] = []
-
+    ) -> Optional[Sequence[SubscriptionResultFuture]]:
+        # Schedule all of the queries (tasks) associated with this tick message
+        # for (asynchronous) execution.
+        # NOTE: There is no backpressure here (tasks will be submitted to the
+        # executor as quickly as they are received from the broker) but there
+        # *is* a concurrency limit on the executor. If a much larger number of
+        # queries are scheduled to execute than there is capacity to execute,
+        # waiting for them to complete during ``flush_batch`` may exceed the
+        # consumer poll timeout (or session timeout during consumer
+        # rebalancing) and cause the entire batch to be have to be replayed.
```

Comment:
This will need to get fixed at the executor level at some point, probably by implementing our own Executor implementation that is backed by a bounded queue, or by feeding the executor through a bounded queue instead of submitting items directly. I forgot that this was one of my major issues with the `concurrent.futures` implementation.

Reply:
So if I get it right, your concern is really the poll timeout, during temporary slowdowns or clickhouse overload right ?
If we regularly have more queries to execute than we can, I don't think backpressure can do much about it since ticks will be still produced at the same rate and we are not willing to skip a tick for load shedding. Is this correct?

Reply:
> So if I get it right, your concern is really the poll timeout, during temporary slowdowns or clickhouse overload right ?

Pretty much. It's possible to enqueue up too many queries (since `process_message` is asynchronous and the queue size is unbounded) that it is not possible to wait for all of them to complete during the the blocking batch flush operation. I think this is more likely during recovery after a period of substantial consumer downtime (bad deploy that gets rolled back, temporary database outage, etc.) than as part of a gradual buildup to a tipping point.

> If we regularly have more queries to execute than we can, I don't think backpressure can do much about it since ticks will be still produced at the same rate and we are not willing to skip a tick for load shedding. Is this correct?

Sort of, yeah. I think the benefit of having backpressure here would be that it decreases the probability of the batch flush blocking long enough that the batch times out and has to be retried, adding load to an already overloaded system. You're right in that it doesn't _fix_ the problem, but it would at least help to avoid an already bad situation becoming worse.

I do think we will need to have some story for load shedding. I wouldn't say we're not willing to skip ticks, but it should be a last resort.